### ISSUE TITLE
Semantics/statefulness of Solver.get_model not clear

### ISSUE THREAD

**Issue by colinmorris** [pysmt/pysmt#366]
This snippet illustrates some aspects of Solvers that I'm finding confusing:

```python
import pysmt.shortcuts as sc
import pysmt.typing as tp
import sys
cmds = sys.argv[1].strip() if len(sys.argv) > 1 else '123'

a = sc.Symbol('A', tp.INT)
b = sc.Symbol('B', tp.INT)
domain = sc.And(
  (1 <= a), (a <= 3),
  (1 <= b), (b <= 5)
)
solver = sc.Solver('msat', incremental=True)

if '1' in cmds:
  print solver.is_sat(domain)
if '2' in cmds:
  solver.add_assertion(domain)
if '3' in cmds:
  print solver.is_sat(domain)
print solver.get_model()
```
(You can pass a string of numbers on the command line to control which of the commands at the bottom are executed. I experimented with a couple different Solver back-ends, and it didn't seem to make a difference.)
### Expected behaviour:
I would think that only `add_assertion` modifies the state of the solver and causes it to 'remember' the constraints in `domain`. As long as 2 is run, this should print out a model like A=1, B=1. If 1/3 are run but not 2, I would expect either an exception, or maybe a model with values for A and B that don't necessarily agree with `domain`
### Actual behavior
'2' alone raises a generic-looking exception:
> pysmt.exceptions.InternalSolverError: no model available

'12' raises the same exception, but succeeds if the 'incremental' option is set to True. '1', '3', '13', '23', and '123' all succeed.

Why does calling `get_model` immediately after `add_assertion` raise an Exception? Maybe `get_model` is only intended to be called after `solve` has been called, and it just happens that `is_sat` calls `solve` as a side effect? If so, I think that precondition should be documented.

It also seems a little odd to me that, unlike `solve`, `get_model` doesn't optionally accept a collection of formulas. I wonder if there's a reason for this?

**Comment by marcogario** [comment]
Thank you very much for taking the time of looking into this. The behavior that you get is (mostly) what we expect. For sure we can improve the documentation, to clarify this.

I'll try to clarify some points here:

* ```get_model``` can be called only after a ```solve``` (or a derived method) has shown that the current assertion stack (+ optional assumptions) is satisfiable. 
* ```solve``` is the best way to interact with the solver. It checks the satisfaction of the stack of assertions. It optionally takes a set of assumptions that are considered together with the assertions in a "temporary" way. A call to ```add_assertion``` modifies the assertion stack, thus any successive call to ```solve``` will take it into account. Assumptions, instead, are removed when you issue a new ```solve```.
* ```is_sat```, ```is_unsat```, ```is_valid```, are helper methods that use assumptions. Functionally, ```is_sat``` is equivalent to ```solve```. The other two are syntactic sugar around ```is_sat```.
* ```get_model``` does not perform solving, for this reason it does not take a set of formulas in input.

```
If 1/3 are run but not 2, I would expect either an exception, or maybe a model with values for A and B that don't necessarily agree with domain
```
Since you are using the domain as an assumption, it is taken into account. The behavior that you are expecting would occur if you were to do ```is_sat()``` without arguments.

```
'12' raises the same exception, but succeeds if the 'incremental' option is set to True. '
```
I think you wanted to say the opposite. If 'incremental' is set to False, this succeeds. This is indeed a bug that was recently introduce. For your '12' execution you should always get the exception.

Finally, please not that the first argument of ```Solver``` is a boolean that indicates whether you want a quantified solver, and it is not the name of the solver. This is in the process of being deprecated, but for now you need to pay attention and do ```Solver(name='msat')``` instead.

In summary, I think we will:
* Clarify the documentation for ```get_model```, specifying that it should be called only after a successful sat-check;
* (Maybe) Explicitely check that ```get_model``` is called after a successful sat check (and introduce a dedicated exception)
* Investigate and fix the bug related to is_sat when incremental=False.

Please, let me know if anything is still unclear.

**Comment by colinmorris** [comment]
Sounds good. Thanks for the thorough explanation!

**Comment by marcogario** [comment]
PR #368
